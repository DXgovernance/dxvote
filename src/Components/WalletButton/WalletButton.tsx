import { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';

import { getChains, injected } from 'provider/connectors';
import { useRpcUrls } from 'provider/providerHooks';
import { Button } from 'old-components/Guilds/common/Button';
import { useWeb3React } from '@web3-react/core';
import WalletModal from 'old-components/Guilds/Web3Modals/WalletModal';
import AddressButton from 'Components/AddressButton/AddressButton';
import { useTransactions } from 'contexts/Guilds';

const WalletButton = () => {
  const { t } = useTranslation();
  const { account, chainId } = useWeb3React();
  const { transactions } = useTransactions();
  const [isWalletModalOpen, setIsWalletModalOpen] = useState(false);

  const [injectedWalletAuthorized, setInjectedWalletAuthorized] =
    useState(false);
  const rpcUrls = useRpcUrls();

  useEffect(() => {
    injected.isAuthorized().then(isAuthorized => {
      setInjectedWalletAuthorized(isAuthorized);
    });
  }, []);

  const toggleWalletModal = () => {
    setIsWalletModalOpen(!isWalletModalOpen);
  };

  const switchNetwork = async chain => {
    const chainIdHex = `0x${chain.id.toString(16)}`;
    try {
      await window.ethereum?.send('wallet_switchEthereumChain', [
        { chainId: chainIdHex },
      ]);
    } catch (e: any) {
      if (e?.code === 4902) {
        window.ethereum?.send('wallet_addEthereumChain', [
          {
            chainId: chainIdHex,
            chainName: chain.displayName,
            nativeCurrency: chain.nativeAsset,
            rpcUrls: [chain.rpcUrl, chain.defaultRpc],
            blockExplorerUrls: [chain.blockExplorer],
          },
        ]);
      }
    }
  };

  function getWalletStatus() {
    if (injectedWalletAuthorized && !account) {
      const chains = getChains(rpcUrls);
      const activeChain =
        chains.find(chain => chain.id === chainId) || chains[0];

      const isMetamask = window.ethereum && window.ethereum.isMetaMask;

      return (
        <Button variant="secondary" onClick={() => switchNetwork(activeChain)}>
          {t('switch')} {isMetamask ? 'MetaMask' : t('wallet')} {t('to')}{' '}
          {activeChain.displayName}
        </Button>
      );
    } else if (account) {
      return (
        <AddressButton
          address={account}
          transactionsCounter={
            transactions?.filter(transaction => !transaction.receipt)?.length ??
            0
          }
          onClick={toggleWalletModal}
        />
      );
    } else {
      return (
        <Button data-testid="connectWalletBtn" onClick={toggleWalletModal}>
          {t('connectWallet')}
        </Button>
      );
    }
  }

  return (
    <>
      {getWalletStatus()}
      <WalletModal isOpen={isWalletModalOpen} onClose={toggleWalletModal} />
    </>
  );
};

export default WalletButton;
