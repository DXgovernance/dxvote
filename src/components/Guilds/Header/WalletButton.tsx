import styled from 'styled-components';
import { observer } from 'mobx-react';
import { isDesktop } from 'react-device-detect';
import WalletModal from 'components/WalletModal';
import { getChains, injected } from 'provider/connectors';
import { useEffect, useState } from 'react';
import { useRpcUrls } from 'provider/providerHooks';
import { useContext } from '../../../contexts';
import { Button, IconButton } from '../common/Button';
import dxIcon from '../../../assets/images/dxdao-icon.svg';

const ButtonIcon = styled.img`
  height: 1.1rem;
  width: 1.1rem;
`;

const Web3Status = observer(() => {
  const {
    context: { modalStore, providerStore },
  } = useContext();
  const [injectedWalletAuthorized, setInjectedWalletAuthorized] =
    useState(false);
  const rpcUrls = useRpcUrls();

  useEffect(() => {
    injected.isAuthorized().then(isAuthorized => {
      setInjectedWalletAuthorized(isAuthorized);
    });
  }, [injected]);

  const { chainId, account } = providerStore.getActiveWeb3React();

  const toggleWalletModal = () => {
    modalStore.toggleWalletModal();
  };

  const switchNetwork = async chain => {
    const chainIdHex = `0x${chain.id.toString(16)}`;
    try {
      await window.ethereum?.send('wallet_switchEthereumChain', [
        { chainId: chainIdHex },
      ]);
    } catch (e: any) {
      if (e?.code == 4902) {
        window.ethereum?.send('wallet_addEthereumChain', [
          {
            chainId: chainIdHex,
            chainName: chain.displayName,
            nativeCurrency: chain.nativeAsset,
            rpcUrls: [chain.rpcUrl, chain.defaultRpc],
            blockExplorerUrls: [chain.blockExplorer],
          },
        ]);
      }
    }
  };

  function getWalletStatus() {
    if (injectedWalletAuthorized && !account) {
      const chains = getChains(rpcUrls);
      const activeChain =
        chains.find(chain => chain.id == chainId) || chains[0];
      const isMetamask = window.ethereum && window.ethereum.isMetaMask;

      return (
        <Button onClick={() => switchNetwork(activeChain)}>
          Switch {isMetamask ? 'MetaMask' : 'Wallet'} to{' '}
          {activeChain.displayName}
        </Button>
      );
    } else if (account) {
      return (
        <IconButton onClick={toggleWalletModal} iconLeft>
          <ButtonIcon src={dxIcon} alt={'Icon'} />{' '}
          {isDesktop && <span>{account}</span>}
        </IconButton>
      );
    } else {
      return <Button onClick={toggleWalletModal}>Connect Wallet</Button>;
    }
  }

  return (
    <>
      {getWalletStatus()}
      <WalletModal />
    </>
  );
});

export default Web3Status;
