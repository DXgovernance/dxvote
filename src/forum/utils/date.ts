import moment from 'moment';

export const formatTime = date => {
  const d = new Date(date);
  return moment(d).format('MMM dd');
};

