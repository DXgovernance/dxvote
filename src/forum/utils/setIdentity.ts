import Identities from 'orbit-db-identity-provider';
import { getResolver } from '@ceramicnetwork/3id-did-resolver';

// Connects the logged in Ceramic user to OrbitDB
export async function setIdentity({ orbitdb, registry }, { ceramic, threeId }) {
  if (orbitdb.identity.type !== 'did') {
    Identities.DIDIdentityProvider.setDIDResolver(getResolver(ceramic));
    const identity = await Identities.createIdentity({
      type: 'DID',
      didProvider: threeId.getDidProvider(),
    });
    orbitdb.identity = identity;
    registry.setIdentity(identity);
    console.log('OrbitDB: DID identity set!', identity);
  }
}
