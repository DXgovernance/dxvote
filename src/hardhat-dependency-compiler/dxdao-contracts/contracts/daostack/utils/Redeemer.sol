// SPDX-License-Identifier: UNLICENSED
pragma solidity >0.0.0;
import 'dxdao-contracts/contracts/daostack/utils/Redeemer.sol';
