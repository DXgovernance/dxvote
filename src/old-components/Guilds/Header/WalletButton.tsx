import { getChains, injected } from 'provider/connectors';
import { useEffect, useState } from 'react';
import { useRpcUrls } from 'provider/providerHooks';
import { Button } from '../common/Button';
import { useWeb3React } from '@web3-react/core';
import WalletModal from '../Web3Modals/WalletModal';
import AddressButton from '../AddressButton';
import { useTransactions } from '../../../contexts/Guilds';

const Web3Status = () => {
  const { account, chainId } = useWeb3React();
  const { transactions } = useTransactions();
  const [isWalletModalOpen, setIsWalletModalOpen] = useState(false);

  const [injectedWalletAuthorized, setInjectedWalletAuthorized] =
    useState(false);
  const rpcUrls = useRpcUrls();

  useEffect(() => {
    injected.isAuthorized().then(isAuthorized => {
      setInjectedWalletAuthorized(isAuthorized);
    });
  }, []);

  const toggleWalletModal = () => {
    setIsWalletModalOpen(!isWalletModalOpen);
  };

  const switchNetwork = async chain => {
    const chainIdHex = `0x${chain.id.toString(16)}`;
    try {
      await window.ethereum?.send('wallet_switchEthereumChain', [
        { chainId: chainIdHex },
      ]);
    } catch (e: any) {
      if (e?.code == 4902) {
        window.ethereum?.send('wallet_addEthereumChain', [
          {
            chainId: chainIdHex,
            chainName: chain.displayName,
            nativeCurrency: chain.nativeAsset,
            rpcUrls: [chain.rpcUrl, chain.defaultRpc],
            blockExplorerUrls: [chain.blockExplorer],
          },
        ]);
      }
    }
  };

  function getWalletStatus() {
    if (injectedWalletAuthorized && !account) {
      const chains = getChains(rpcUrls);
      const activeChain =
        chains.find(chain => chain.id == chainId) || chains[0];
      const isMetamask = window.ethereum && window.ethereum.isMetaMask;

      return (
        <Button variant="secondary" onClick={() => switchNetwork(activeChain)}>
          Switch {isMetamask ? 'MetaMask' : 'Wallet'} to{' '}
          {activeChain.displayName}
        </Button>
      );
    } else if (account) {
      return (
        <AddressButton
          address={account}
          transactionsCounter={
            transactions.filter(transaction => !transaction.receipt).length
          }
          onClick={toggleWalletModal}
        />
      );
    } else {
      return (
        <Button data-testId="connectWalletBtn" onClick={toggleWalletModal}>
          Connect Wallet
        </Button>
      );
    }
  }

  return (
    <>
      {getWalletStatus()}
      <WalletModal isOpen={isWalletModalOpen} onClose={toggleWalletModal} />
    </>
  );
};

export default Web3Status;
