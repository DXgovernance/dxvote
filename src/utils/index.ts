export * from './address';
export * from './bignumber';
export * from './cache';
export * from './constants';
export * from './date';
export * from './enums';
export * from './helperHooks';
export * from './helpers';
export * from './number';
export * from './permissions';
export * from './proposals';
export * from './scheme';
export * from './transaction';
export * from './encodingCalls';
export * from './signature';
export * from './array';
export * from './votes';
