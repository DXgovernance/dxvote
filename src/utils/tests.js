import { ThemeProvider } from 'styled-components';
import { render as rtlRender } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { GuildsDarkTheme } from 'Components/theme';

export const renderWithRoutes = (ui, options = {}) => {
  //wrap with MemoryRouter
  return rtlRender(ui, { wrapper: MemoryRouter, ...options });
};

// we are mokcing testing-library/render with our own render
export const render = (children, options = {}) => {
  //wrap with Theme
  return renderWithRoutes(
    <ThemeProvider theme={GuildsDarkTheme}> {children} </ThemeProvider>,
    options
  );
};
