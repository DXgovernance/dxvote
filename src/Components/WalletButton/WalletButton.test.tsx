import WalletButton from './WalletButton';
import { render } from 'utils/tests';
import { useWeb3React } from '@web3-react/core';
import { ANY_ADDRESS, shortenAddress } from 'utils';

jest.mock('hooks/Guilds/ether-swr/ens/useENSAvatar', () => ({
  __esModule: true,
  default: () => ({
    avatarUri: 'test',
    imageUrl: 'test',
    ensName: 'test.eth',
  }),
}));

jest.mock('contexts/index', () => jest.fn());
jest.mock('contexts/Guilds', () => {
  return {
    useTransactions: () => {
      return [];
    },
  };
});

jest.mock('provider/providerHooks', () => {
  return {
    useRpcUrls: () => {
      return {
        '1': 'https://mainnet.infura.io/v3/dummy',
      };
    },
  };
});

jest.mock('@web3-react/core', () => ({
  useWeb3React: jest.fn(),
}));

const mockedUseWeb3React = useWeb3React as jest.Mock;

describe('WalletButton', () => {
  it('Should match snapshot and display connect wallet', () => {
    mockedUseWeb3React.mockImplementation(() => ({
      account: null,
      chainId: null,
    }));
    const { container, getByText } = render(<WalletButton />);
    expect(container).toMatchSnapshot();
    expect(getByText('connectWallet')).toBeInTheDocument();
  });
  it('Should match snapshot and display connected address', () => {
    mockedUseWeb3React.mockImplementation(() => ({
      account: ANY_ADDRESS,
      chainId: 1,
    }));
    const { container, getByText } = render(<WalletButton />);
    expect(container).toMatchSnapshot();
    expect(getByText(shortenAddress(ANY_ADDRESS))).toBeInTheDocument();
  });
});
