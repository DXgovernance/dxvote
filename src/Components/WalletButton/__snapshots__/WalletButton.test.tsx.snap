// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletButton Should match snapshot and display connect wallet 1`] = `
<div>
   
  <button
    class="sc-bdfBQB cnzpap"
    data-testid="connectWalletBtn"
  >
    connectWallet
  </button>
   
</div>
`;

exports[`WalletButton Should match snapshot and display connected address 1`] = `
<div>
   
  <button
    class="sc-bdfBQB sc-gsTEea sc-cOajNj cAORUT hqMJhH hmnXkE"
  >
    <span
      class="sc-bBrNTY kkgJGY"
    >
      <img
        alt="Avatar"
        class="sc-dlfnuX sc-kfzBvY hZGubn btMxwF"
        data-testid="avatar"
        size="24"
        src="test"
      />
    </span>
    <span
      class="sc-Azgjq jkyMEq"
    >
      0xaAaA...aaAa
    </span>
  </button>
   
</div>
`;
